### PR TITLE
`driver`: refactor the tests and re-add commented tests

### PR DESCRIPTION
**Pull request by cowfee**
Fixes https://github.com/cowprotocol/services/issues/1527 and https://github.com/cowprotocol/services/issues/1306.

Follow-up to https://github.com/cowprotocol/services/pull/1526.

Refactors all tests to use the new test setup framework. Makes the tests way shorter and easier to write.

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]
The code and test structure looks good. I haven't looked at what exactly is being tested closely.

File: crates/driver/src/domain/quote.rs [hunk 1]
```diff
@@ -143,8 +147,8 @@ impl Order {
                 token: self.tokens.sell,
             },
             // Note that we intentionally do not use [`eth::U256::max_value()`]
-            // as an order with this would cause overflows with the Smart
-            // Contract, so buy orders requiring excessively large sell amounts
+            // as an order with this would cause overflows with the smart
+            // contract, so buy orders requiring excessively large sell amounts
```
Comment (nlordell) [inline crates/driver/src/domain/quote.rs:151]: I think in theory, "Smart Contract" should be capitalized in the context of Ethereum Smart Contracts. At least this is how I thought the Ethereum foundation capitalized it.

Comment (nlordell) [reply crates/driver/src/domain/quote.rs:151]: **Edit**: I stand corrected, it is definitely "smart contract" and I've been doing it wrong for a long time 🙈

**Review by nlordell — APPROVED** [review]
Looks great! I carefully reviewed the domain and infra changes and they look good.

Looked at a test, and it is much nicer to write integration tests now!

File: crates/driver/src/tests/setup/solver.rs [hunk 2]
```diff
@@ -1,63 +1,235 @@
...
+                "buyAmount": match quote.order.side {
+                    order::Side::Sell if config.quote => "1".to_owned(),
+                    _ => quote.buy_amount().to_string(),
+                },
+                "feeAmount": quote.order.user_fee.to_string(),
+                "kind": match quote.order.side {
+                    order::Side::Sell => "sell",
+                    order::Side::Buy => "buy",
+                },
+                "partiallyFillable": matches!(quote.order.partial, order::Partial::Yes { .. }),
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/solver.rs:70]: nit: I believe in case of `config.quote` this needs to be `false`. A quote for a partial fill is pretty much useless for our purposes. Plus you already correctly set `class: market` in this case which is incompatible with `partiallFillable: true`.

File: crates/driver/src/tests/setup/solver.rs [hunk 3]
```diff
@@ -1,63 +1,235 @@
...
-/// Set up an HTTP server exposing a solver API and acting as a solver mock.
-pub async fn setup(config: Config) -> Solver {
-    let state = Arc::new(Mutex::new(config.solve.clone()));
-    let app = axum::Router::new()
+impl Solver {
+    /// Set up an HTTP server exposing a solver API and acting as a solver mock.
+    pub async fn new(config: Config<'_>) -> Self {
+        let mut solutions_json = Vec::new();
+        let mut orders_json = Vec::new();
+        for quote in config.quotes {
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/solver.rs:40]: If I understood the `Test` code correctly `quote` tests are mutually exclusive with `solve` tests so it would make sense to me to either build responses for `quotes` or `solutions` here.

File: crates/driver/src/tests/setup/mod.rs [hunk 4]
```diff
@@ -1,12 +1,988 @@
...
+    pub internalize: bool,
+    pub side: order::Side,
+    pub partial: order::Partial,
+    pub valid_for: util::Timestamp,
+    pub kind: order::Kind,
+
+    // TODO For now I'll always set these to zero. But I think they should be tested as well.
+    // Figure out what (if anything) would constitute meaningful tests for these values.
+    pub user_fee: eth::U256,
+    pub solver_fee: eth::U256,
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/mod.rs:97]: The `solver_fee` should only ever affect the `score` of the solution. So you could have a test with 2 identical orders but one has a bigger `solver_fee`. Then the one with the bigger `solver_fee` should also have a higher score.

And for the `user_fee` you could assert that the `user_fee` remains in the settlement contract after settling an order with non-zero `user_fee`.

**Review by MartinquaXD — COMMENTED** [review]
The new tests look very nice.
Just a bunch of nits and some technicalities which probably don't really need to be addressed.

File: crates/driver/src/infra/simulator/mod.rs [hunk 5]
```diff
@@ -26,16 +30,20 @@ impl Simulator {
...
-            disable_access_lists: true,
-            ..self
-        }
+    pub fn disable_access_lists(&mut self) {
+        self.disable_access_lists = true;
+    }
+
+    /// Disable gas simulation. Useful for testing, but shouldn't be used in
+    /// production since it will cause the driver to return invalid scores.
+    pub fn disable_gas(&mut self, fixed_gas: eth::Gas) {
```
[the remark this answers is not in the document]
Comment (cowfee) [reply crates/driver/src/infra/simulator/mod.rs:45]: I think this _might_ also be useful when manually running. The main reason is because nodes tend to give more details in the response of `eth_sendTransaction` than they do in the response of `eth_call`. In particular I've noticed that the `eth_call` on geth doesn't return the revert message. Also, it's easier to inspect and debug published txs.

File: crates/driver/src/tests/setup/mod.rs [hunk 6]
```diff
@@ -1,12 +1,988 @@
...
+            .transaction(TransactionId::Block(
+                BlockId::Number(ethcontract::BlockNumber::Latest),
+                0.into(),
+            ))
+            .await
+            .unwrap()
+            .unwrap();
+        let input = tx.input.0;
+        let len = input.len();
+        let tx_solution_id = u64::from_be_bytes((&input[len - 8..]).try_into().unwrap());
```
[the remark this answers is not in the document]
Comment (cowfee) [reply crates/driver/src/tests/setup/mod.rs:939]: Actually, nowadays this _is_ guaranteed. Since the driver submits the settlement, there has to be a way to the autopilot to recognize it. This happens via the settlement ID. Here's the [related domain type](https://github.com/cowprotocol/services/blob/dec4abfb0f540344c5337e125e7f210d6826d85d/crates/driver/src/domain/competition/solution/settlement.rs#L326-L332).

[the remark this answers is not in the document]
Comment (cowfee) [reply crates/driver/src/tests/setup/mod.rs:939]: Do you think I should improve the doc comment to mention that it's encoded _at the end_ of the calldata?

[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/driver/src/tests/setup/mod.rs:939]: I know that the driver already does this at the moment. My point was if we were to change that logic and introduce a bug which for example adds 1 byte less this code would not pick the correct bytes as the id to compare against. In that scenario it would make the last byte of the actual call data part of the id to compare against and **might not** detect that regression.
However, the chance that miraculously all these ids match by accident to please the CI pipeline is probably very low.

File: crates/driver/src/tests/setup/blockchain.rs [hunk 7]
```diff
@@ -292,7 +286,7 @@ impl Blockchain {
         // Create the tokens needed by the pools.
         let mut tokens = HashMap::new();
         for pool in config.pools.iter() {
-            if !tokens.contains_key(pool.reserve_a.token) {
+            if pool.reserve_a.token != "WETH" && !tokens.contains_key(pool.reserve_a.token) {
```
[the remark this answers is not in the document]
Comment (cowfee) [reply crates/driver/src/tests/setup/blockchain.rs:289]: This code doesn't mint, it deploys the tokens. WETH is already deployed above because it's "special". Edge cases are a PITA...

File: crates/driver/src/tests/setup/mod.rs [hunk 8]
```diff
@@ -1,12 +1,988 @@
...
+    }
+}
+
+/// An order which buys ETH.
+pub fn eth_order() -> Order {
+    Order {
+        name: "ETH order",
+        amount: ETH_ORDER_AMOUNT.into(),
+        sell_token: "A",
+        buy_token: "WETH",
```
[the remark this answers is not in the document]
Comment (cowfee) [reply crates/driver/src/tests/setup/mod.rs:391]: This was a bug, thanks for catching it...

File: crates/driver/src/tests/setup/solver.rs (hunk 3 above)
Comment (cowfee) [reply crates/driver/src/tests/setup/solver.rs:40]: This is one of the more messy parts of the setup framework. These _quotes_ are `setup::blockchain::Quote`s, which simply represent the achievable prices. They have nothing to do with the driver quoting process. :grimacing:
